src/Pact/Analyze/Check.hs

```diff
@@ -742,46 +799,174 @@ verifyFunctionProps
-    Right (TopFun FDefun {_fInfo, _fArgs, _fBody} _) ->
-      for props $
-        verifyFunctionProperty modName funName _fInfo tables caps _fArgs _fBody
-          checkType
-    Right _ ->
-      pure []
+verifyFunctionProps modName tables caps funName props checkType = \case
+  FDefun {_fInfo, _fArgs, _fBody} ->
+    for props $
+      verifyFunctionProperty modName funName _fInfo tables caps _fArgs _fBody
+        checkType
+  _ ->
+    pure []
 
 verifyFunctionInvariants
   :: ModuleName
   -> [Table]
   -> [Capability]
-  -> Ref
+  -> TopLevel Node
   -> Text
   -> CheckableType
   -> IO (Either CheckFailure (TableMap [CheckResult]))
-verifyFunctionInvariants modName tables caps ref funName checkType = do
-  eToplevel <- typecheck ref
-  case eToplevel of
-    Left failure ->
-      pure $ Left failure
-    Right (TopFun FDefun {_fInfo, _fArgs, _fBody} _) ->
+verifyFunctionInvariants modName tables caps topLevel funName checkType =
+  case topLevel of
+    TopFun FDefun {_fInfo, _fArgs, _fBody} _ ->
       verifyFunctionInvariants' modName funName _fInfo tables caps _fArgs
         _fBody checkType
-    Right _ ->
+    _ ->
       pure $ Right $ TableMap Map.empty
 
 -- TODO: use from Control.Monad.Except when on mtl 2.2.2
 liftEither :: MonadError e m => Either e a -> m a
 liftEither = either throwError return
 
+getConsts
+  :: HM.HashMap Text Ref
+  -> ExceptT VerificationFailure IO (HM.HashMap Text EProp)
+getConsts defconstRefs = do
+
+  (consts :: HM.HashMap Text (AST Node)) <- ifoldrM
+    (\name ref accum -> do
+      maybeConst <- lift $ runTC 0 False $ typecheckTopLevel ref
+      pure $ case maybeConst of
+        (TopConst _info _qualifiedName _type val _doc, _tcState)
+          -> accum & at name ?~ val
+        _ -> error "invariant failure: anything but a const is unexpected here"
+    )
+    HM.empty
+    defconstRefs
+
+  let constToProp :: ETerm -> Except VerificationFailure EProp
+      constToProp tm = case constantToProp tm of
+        Right prop -> pure prop
+        Left msg   -> throwError $ FailedConstTranslation msg
+
+      translateNodeNoGraph'
+        = withExceptT translateToVerificationFailure . translateNodeNoGraph
+
+  hoist generalize $
+    traverseOf each (constToProp <=< translateNodeNoGraph') consts
+
+getStepChecks
+  :: [Table]
+  -> HM.HashMap Text EProp
+  -> HM.HashMap Text (DefinedProperty (Exp Info))
+  -> (CheckableType -> Text -> [Located Check] -> Fun Node -> IO [CheckResult])
+  -> HM.HashMap Text Ref
+  -> ExceptT VerificationFailure IO (HM.HashMap (Text, Int) [CheckResult])
+getStepChecks tables consts propDefs verifyFunProps defpactRefs = do
+
+  (steps :: HM.HashMap (Text, Int)
+    ((AST Node, [Named Node], Info), Pact.FunType TC.UserType))
+    <- ifoldrM
+    (\name ref accum -> do
+      maybeDef <- lift $ runTC 0 False $ typecheckTopLevel ref
+      pure $ case maybeDef of
+        (TopFun (FDefun info _ _ Defpact funType args steps) _meta, _)
+          -> ifoldl
```

Comment:
It seems a bit unfortunate to have `ifoldl` inside `ifoldrM`. The idea is that for each pact, we're looking through each step.

Reply:
That seems fine to me